pkg/util/fsync/fsync.go

```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package fsync
+
+import (
+	"os"
+	"path"
+
+	"github.com/grafana/dskit/multierror"
+)
+
+// Run execute writeFunc to write data into passed file and then execute
+// fsync operation to make sure the file and its content are stored atomically.
+func Run(file *os.File, writeFunc func(f *os.File) (int, error)) error {
```

Comment:
Rather than taking `writeFunc`, could this function take a string or a byte slice to write to the file instead?

Reply:
Can we do union of string or []byte? Also the function that it uses are different. `Write` vs `WriteString`. Let me see again.

Reply:
Perhaps `[]byte` is the simplest choice? Converting from `string` to `[]byte` isn't too difficult for consumers that would prefer to work with strings.

Reply:
why not go with the standard `io.Reader`? I think this allows for writing larger files and is a standard in go. `os.File` also implements `io.ReaderFrom` which takes care of tracking `io.EOF` and making sure everything from the reader is written to the file.